### ISSUE TITLE
Add a chapter on behaviour classification

### ISSUE THREAD

**Issue by niksirbi** [neuroinformatics-unit/course-animals-in-motion#16]
Several participants expressed the desire to also cover behaviour classification/segmentation.
Perhaps we should provide a basic overview of available approaches/tools + a hands-on tutorial on one specific tool, similar to what we have done for pose estimation with SLEAP.

That said, delivering this chapter on top of all existing materials would probably necessitate a course duration of > 2 days.

**Comment by niksirbi** [comment]
[This Transmitter article](https://www.thetransmitter.org/behavioral-neuroscience/beyond-the-algorithmic-oracle-rethinking-machine-learning-in-behavioral-neuroscience) makes some good points about action segmentation.

**Comment by niksirbi** [comment]
This movement-BORIS example could come in handy for that: https://github.com/neuroinformatics-unit/movement/pull/865

We could apply sth like that on CalMS21 data, specifically:

- Use BORIS to annotate behaviours (sth simple like mounting, or the 3-4 behaviours defined in the CalMS21 dataset)
- Import the labels into Python
- Train a supervised classifiers (XGBoost or similar)
- Predict behavioural labels on a new video and attach them to a `movement` dataset

**Comment by niksirbi** [comment]
I will move ahead and draft a practical chapter based on what I've outlined in my [previous comment](https://github.com/neuroinformatics-unit/course-animals-in-motion/issues/16#issuecomment-4214124014).

The practical chapter will be preceded by a 'theory' chapter introducing the concepts of supervised/unsupervised/semi-supervised behaviour classification.

Pros of sticking with the CalMS21 data:
- Learners have been already introduced to it
- It helpfully comes with ground-truth behaviour annotations, on top of the keypoint annotations

Pros of teaching [BORIS](https://www.boris.unito.it):
- Genuinely useful mature open-source tool, useful for all kinds of behaviour annotation scenarios
- Works with any species, in any environment
- Holly has already given us a great [movement example](https://github.com/neuroinformatics-unit/course-animals-in-motion/issues/16#issuecomment-4214124014). Adapting that to the CalMS21 case and writing it as a chapter for the handbook should bot be too hard

Cons of teaching BORIS:
- There are some [complications with getting it to run on MacOS](https://www.boris.unito.it/download_mac/), and I expect many learners to have Macbooks
- It will take some time to go through a hands-on tutorial

### Question

Should we add BORIS to the pre-requisites, and just deal with the MacOS problems?
Or would we rather not use BORIS in the course and use the already annotated CalMS21 frames as a starting point for training. 
The latter choice is much easier, but it's kind of unsatisfying, because learners will have no easy to way to label their own data.
To help us decide, I will try to see if BORIS works on my MacOS. @lochhh would be helpful if you could try installing it on Windows and report if it installs/launches seamlessly.

We could also teach [JABS](https://github.com/KumarLabJax/JABS-behavior-classifier) instead of BORIS, but that's mouse specific.

Any thoughts @neuroinformatics-unit/behaviour ?

**Comment by lochhh** [comment]
happy to try BORIS first as we expect people to work with different kinds of animals, where JABS would be less useful.

**Comment by niksirbi** [comment]
Today, I also explored the possibility of using [JABS](https://github.com/KumarLabJax/JABS-behavior-classifier), because I really like their GUI-based active learning workflow (especially for teaching purposes), but I don't think it's appropriate for the "Animals in Motion" crowd yet, becasue:
- poses have to be in a very specific format (that `movement` doesn't yet support)
- It's only for mice and requires 12 keypoints.
- The hand-crafted features (see Table S4 in [paper supplement](https://elifesciences.org/reviewed-preprints/107259v2#s5)) are derived from these 12 keypoints.

In summary, JABS would be good for demonstrating the 'concept' of supervised behaviour classification, but most Animals in Motion learners wouldn't be able to apply that to their own data (which is the ethos of our course). 

I also looked into MARS, and it has the same drawbacks: mouse-only, specific keypoints, hand-crafted features assuming those keypoints only.

**Comment by niksirbi** [comment]
As an aside:

It would be great to have a GUI-based app with the following features:
- Species and pose format agnostic (e.g. using `movement` to load any tracking data)
- JABS-like active learning interface for interactive behaviour labelling (possibly assisted by change-point detection, with ethograph-like 'snapping' to changepoints).
- Some flexibility in what features are being computed (could also use `movement` under the hood).

Such an app doesn't exist, but perhaps it's something we should think about as a future project idea.

**Comment by sfmig** [comment]
I chatted to one of the FERAL developers yesterday, they have a colab that maybe we can adapt. Maybe we should give it a go? It seems to perform well in the sample datasets and the pipeline seems simple enough. I leave the link in case it is useful: https://github.com/Skovorp/feral#quickest-start-google-colab.

They also support loading BORIS labels (they have a BORIS plugin too)

**Comment by niksirbi** [comment]
> I chatted to one of the FERAL developers yesterday, they have a colab that maybe we can adapt. Maybe we should give it a go? It seems to perform well in the sample datasets and the pipeline seems simple enough. I leave the link in case it is useful: [Skovorp/feral#quickest-start-google-colab](https://github.com/Skovorp/feral?rgh-link-date=2026-07-30T17%3A19%3A28.000Z#quickest-start-google-colab).

I think it's worth exploring/demo-ing FERAL during the collaborations days, but I'm less sure about teaching it as part of the behaviour classification chapter.

The simple pipeline (which should work on any species) and the support for BORIS labels is definitely a plus!

On the downside, FERAL primarily runs through colab, and requires large GPUs (probably because the pipeline is quite resource intensive?).  Options for running it locally are restricted, since most learners won't have Linux machines (the recommended OS) or large enough GPUs. Colab is not so much a problem from the teaching perspective (though we would have to tell everyone to get a Colab pro account), but it goes somewhat against our approach of teaching tools that people can 'go home with'.

Moreover, since we would have spent the previous 2 days of the course talking about pose tracking, a keypoint-based classification approach would fit in more in the course, and would likely be less resource-intensive.

Anyhow, I'm keen to hear counter-arguments and open to being convinced in favour of FERAL.

In the meantime I will start drafting a chapter with a BORIS tutorial. No matter which supervised behaviour classification approach we go for (we can discuss that next week), annotating behaviours will be a necessary starting point for all.

**Comment by lochhh** [comment]
I think we should definitely try it out. If time allows and one of us manages to try it before OSSS, even if we don't teach it (this year), we can frame it in the course like "we've also tried FERAL, JABS..." and share our thoughts on these.

**Comment by sfmig** [comment]
The developer also told me that now they have a FERAL "light" version that can run on 8gb gpu, which is more reasonable. I didn't find clear references to it in the website or repo so not sure it is live.
